content/specs/repository.md

```diff
@@ -0,0 +1,148 @@
+
+In real-world use, it is expected that individual repositories will contain anywhere from dozens to millions of records.
+
+
+## Repo Data Structure (v2)
+
+This describes version `2` of the repository binary format. Version `1` had a different MST fanout configuration, and an incompatible schema for commits and repository metadata.
+
+At a high level, a repository is a key/value mapping where the keys are path names (as strings) and the values are records (DAG-CBOR objects).
+
+A "Merkle Search Tree" (MST) is used to store this mapping. This content-addressed deterministic data structure stores data in key-sorted order. It is reasonably efficient for key lookups, key range scans, and appends (assuming sorted record paths). The properties of MSTs in general are described in this academic publication:
+
+> Alex Auvolat, François Taïani. Merkle Search Trees: Efficient State-Based CRDTs in Open Networks. SRDS 2019 - 38th IEEE International Symposium on Reliable Distributed Systems, Oct 2019, Lyon, France. pp.1-10, ff10.1109/SRDS.2019.00032
+
+The specific details of the MST as used in atproto repositories are described below.
+
+Repo paths are strings, while MST keys are byte arrays. Neither may be empty (zero-length). While repo path strings are currently limited to a subset of ASCII (making encoding a no-op), the encoding is specified as UTF-8.
+
+Repo paths currently have a fixed structure of `<collection>/<record-key>`, which translates to a valid, normalized NSID, followed by a `/`, followed by a valid Record Key. The path should not start with a leading `/`, and should always have exactly two path segments. The ASCII characters allowed in the entire path string are currently: letters (`A-Za-z`), digits (`0-9`), slash (`/`), period (`.`), hyphen (`-`), underscore (`_`), and tilde (`~`).
+
+Note that repo paths for all records in the same collection are sorted together in the MST, making enumeration (via key scan) and export efficient. Additionally, the TID Record Key scheme was intentionally selected to provide chronological sorting of MST keys within the scope of a collection. Appends are more efficient than random insertions/mutations within the tree, and when enumerating records within a collection they will be in chronological order (assuming that TID generation was done "correctly", which can not be relied on in general).
+
+
+### Commit Objects
+
+The top-level data object in a repository is a signed Commit. The IPLD schema fields are:
+
+- `did` (string, required): the account DID associated with the repo, in strictly normalized form (eg, lowercase as appropriate)
+- `version` (integer, required): fixed value of `2` for this repo format version
+- `prev` (CID link, nullable): an optional pointer (by hash) to the most recent previous commit in this repository. any commit without a "prev" pointer, including the first commit, is considered a "rebase"
```

Comment:
in the subscription stream for instance, we actually don't consider the first commit a "rebase"

to me, rebase describes the operation of deleting some historical data not a commit itself with no prev